File: src/Asambleas.Application/Security/Roles.cs
namespace Asambleas.Application.Security;

public static class Roles
{
    public const string PlatformAdmin = "PlatformAdmin";
    public const string TenantAdmin = "TenantAdmin";
    public const string PHAdmin = "PHAdmin";
    public const string AssemblyPresident = "AssemblyPresident";
    public const string AssemblySecretary = "AssemblySecretary";
    public const string AssemblyOperator = "AssemblyOperator";
    public const string Owner = "Owner";
    public const string Auditor = "Auditor";

    public static IReadOnlyList<string> All { get; } =
    [
        PlatformAdmin,
        TenantAdmin,
        PHAdmin,
        AssemblyPresident,
        AssemblySecretary,
        AssemblyOperator,
        Owner,
        Auditor
    ];
}

File: src/Asambleas.Web/wwwroot/js/modules/roles.js
import { hasPermission } from "./auth.js";

const OPERATOR_ROLES = new Set([
  "AssemblyPresident",
  "AssemblySecretary",
  "AssemblyOperator",
  "PHAdmin",
  "Moderator",
  "TenantAdmin",
  "PlatformAdmin"
]);

/**
 * Resolve viewer role for layout. Prefer API viewerRole; else derive from roles/permissions.
 * @returns {"Operator"|"Owner"}
 */
export function resolveViewerRole(user, roomState = null) {
  const fromApi = roomState?.viewerRole || roomState?.viewer?.role;
  if (fromApi === "Operator" || fromApi === "Owner") {
    return fromApi;
  }

  const roles = user?.roles || [];
  if (roles.some((r) => OPERATOR_ROLES.has(r))) {
    return "Operator";
  }

  if (
    hasPermission(user, "assembly:start") ||
    hasPermission(user, "meeting:moderate") ||
    hasPermission(user, "vote:open")
  ) {
    return "Operator";
  }

  return "Owner";
}

export function isOperator(user, roomState = null) {
  return resolveViewerRole(user, roomState) === "Operator";
}

/**
 * Pure owner participant (portal:self / vote:cast) without PH or assembly administration.
 * These users must never see the administrative shell (Panel, Comunicaciones, Nueva asamblea, …).
 */
export function isOwnerPortalUser(user) {
  if (!user) return false;
  if (isOperator(user) || hasPermission(user, "ph:manage") || hasPermission(user, "assembly:manage")) {
    return false;
  }
  if (hasPermission(user, "assembly:schedule") || hasPermission(user, "assembly:reschedule")) {
    return false;
  }
  return hasPermission(user, "portal:self") || hasPermission(user, "vote:cast");
}
